File: examples/python-examples/MeshICP.dox.py
import meshlib.mrmeshpy as mrmeshpy

# Load meshes
meshFloating = mrmeshpy.loadMesh("meshA.stl")
meshFixed = mrmeshpy.loadMesh("meshB.stl")

# Prepare ICP parameters
diagonal = meshFixed.getBoundingBox().diagonal()
icp_sampling_voxel_size = diagonal * 0.01  # To sample points from object
icp_params = mrmeshpy.ICPProperties()
icp_params.distThresholdSq = (diagonal * 0.1) ** 2  # Use points pairs with maximum distance specified
icp_params.exitVal = diagonal * 0.003  # Stop when this distance reached

# Calculate transformation
icp = mrmeshpy.ICP(meshFloating, meshFixed,
                   mrmeshpy.AffineXf3f(), mrmeshpy.AffineXf3f(),
                   icp_sampling_voxel_size)
icp.setParams(icp_params)
xf = icp.calculateTransformation()

# Transform floating mesh
meshFloating.transform(xf)

# Output information string
print(icp.getLastICPInfo())

# Save result
mrmeshpy.saveMesh(meshFloating, "meshA_icp.stl")

File: examples/cpp-examples/MeshICP.dox.cpp
#include <MRMesh/MRBox.h>
#include <MRMesh/MRICP.h>
#include <MRMesh/MRMesh.h>
#include <MRMesh/MRMeshLoad.h>
#include <MRMesh/MRMeshSave.h>

#include <iostream>

int main()
{
    // Load meshes
    auto meshFloatingRes = MR::MeshLoad::fromAnySupportedFormat( "meshA.stl" );
    if ( !meshFloatingRes.has_value() )
    {
        std::cerr << meshFloatingRes.error() << std::endl;
        return 1;
    }
    MR::Mesh& meshFloating = *meshFloatingRes;

    auto meshFixedRes = MR::MeshLoad::fromAnySupportedFormat( "meshB.stl" );
    if ( !meshFixedRes.has_value() )
    {
        std::cerr << meshFixedRes.error() << std::endl;
        return 1;
    }
    MR::Mesh& meshFixed = *meshFixedRes;

//! [0]
    // Prepare ICP parameters
    float diagonal = meshFixed.getBoundingBox().diagonal();
    float icpSamplingVoxelSize = diagonal * 0.01f; // To sample points from object
    MR::ICPProperties icpParams;
    icpParams.distThresholdSq = MR::sqr( diagonal * 0.1f ); // Use points pairs with maximum distance specified
    icpParams.exitVal = diagonal * 0.003f; // Stop when distance reached

    // Calculate transformation
    MR::ICP icp(
        MR::MeshOrPoints{ MR::MeshPart{ meshFloating } },
        MR::MeshOrPoints{ MR::MeshPart{ meshFixed } },
        MR::AffineXf3f(), MR::AffineXf3f(),
        icpSamplingVoxelSize );
    icp.setParams( icpParams );
    MR::AffineXf3f xf = icp.calculateTransformation();

    // Transform floating mesh
    meshFloating.transform( xf );
//! [0]

    // Output information string
    std::string info = icp.getStatusInfo();
    std::cerr << info << std::endl;

    // Save result
    if ( auto saveRes = MR::MeshSave::toAnySupportedFormat( meshFloating, "meshA_icp.stl" ); !saveRes )
    {
        std::cerr << saveRes.error() << std::endl;
        return 1;
    }
}
